src/crimson/os/seastore/seastore_types.h

```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
   }
 };
 
+// internal segment id type of segment_id_t, basically
+// this is a unsigned int with the top "device_id_len"
+// bits representing the id of the device on which the
+// segment resides
+using internal_segment_id_t = uint32_t;
+// type of device id
+using device_id_t = uint8_t;
+// order of device_id_t
+constexpr uint16_t device_id_len = 4;
+// maximum devices supported
+constexpr uint16_t max_devices = 1 << device_id_len;
+
+// segment ids without a device id encapsulated
+using device_segment_id_t = internal_segment_id_t;
+
+// mask for segment manager id
+constexpr internal_segment_id_t SM_ID_MASK =
+  0xF << (std::numeric_limits<internal_segment_id_t>::digits - device_id_len);
+// default internal segment id
+constexpr internal_segment_id_t DEFAULT_INTERNAL_SEG_ID =
+  std::numeric_limits<internal_segment_id_t>::max() - 1;
+
 // Identifies segment location on disk, see SegmentManager,
-using segment_id_t = uint32_t;
+struct segment_id_t {
+  internal_segment_id_t segment = DEFAULT_INTERNAL_SEG_ID;
+
+  segment_id_t() = default;
+  constexpr segment_id_t(internal_segment_id_t segment)
+    : segment(segment) {}
+  segment_id_t(device_id_t id, internal_segment_id_t segment)
+    : segment(add_device_id(segment, id)) {
+    // only lower 4 bits are effective, and we have to reserve 0x0F for
+    // special XXX_SEG_IDs
+    assert(id < 15);
+  }
+
+  [[gnu::always_inline]]
+  device_id_t device_id() const {
+    return get_device_id(segment);
+  }
+
+  [[gnu::always_inline]]
+  internal_segment_id_t internal_segment_id() const {
+    return strip_device_id(segment);
+  }
+
+  bool operator==(const segment_id_t& other) const {
+    return segment == other.segment;
+  }
+  bool operator!=(const segment_id_t& other) const {
+    return segment != other.segment;
+  }
+  bool operator<(const segment_id_t& other) const {
+    return segment < other.segment;
+  }
+  bool operator<=(const segment_id_t& other) const {
+    return segment <= other.segment;
+  }
+  bool operator>(const segment_id_t& other) const {
+    return segment > other.segment;
+  }
+  bool operator>=(const segment_id_t& other) const {
+    return segment >= other.segment;
+  }
+  segment_id_t& operator++() {
```

Comment:
Are there any users that actually increment and decrement this type without converting to device_segment_id_t first?  Should we even allow that?

Reply:
src/test/crimson/seastore/test_seastore_cache.cc:49 and src/test/crimson/seastore/test_seastore_journal.cc:91 are using these operators

Reply:
I don't think adding an integer type to segment_id_t has an obvious meaning (to a reader, it could increment the device, the segment, or both if at the end of a device).  This implementation in particular can cause the segment_id_t to no longer refer to an actual segment.  No user outside of SegmentManager implementations (which will be using device_segment_id_t) or special cases (like these tests where we are basically emulating a SegmentManager) should be manipulating the segment value, so exposing the ability to increment it through the interface invites misuse.

I think you should make those two users (as well as test_btree_lba_manager.cc:59) explicitly construct new segment_id_t instances and remove these operators.

```
--- a/src/test/crimson/seastore/test_btree_lba_manager.cc
+++ b/src/test/crimson/seastore/test_btree_lba_manager.cc
@@ -51,9 +51,14 @@ struct btree_lba_manager_test :
 
   segment_id_t next = 0;
   get_segment_ret get_segment(device_id_t id = 0) final {
+    auto ret = next;
+    next = segment_id_t{
+      next.device_id(),
+      next.internal_segment_id() + 1
+    };
     return get_segment_ret(
       get_segment_ertr::ready_future_marker{},
-      next++);
+      ret);
   }
 
   void set_segment_written_to(paddr_t addr) final {}
diff --git a/src/test/crimson/seastore/test_seastore_cache.cc b/src/test/crimson/seastore/test_seastore_cache.cc
index 1b131869fa8..ceef5a0ae2b 100644
--- a/src/test/crimson/seastore/test_seastore_cache.cc
+++ b/src/test/crimson/seastore/test_seastore_cache.cc
@@ -46,7 +46,11 @@ struct cache_test_t : public seastar_test_suite_t {
                segment_manager->get_segment_size());
     if (current.offset + (segment_off_t)bl.length() >
        segment_manager->get_segment_size())
-      current = paddr_t{current.segment + 1, 0};
+      current = paddr_t{
+       segment_id_t{
+         current.segment.device_id(),
+         current.segment.internal_segment_id() + 1
+       }, 0};
 
     auto prev = current;
     current.offset += bl.length();
```

Reply:
will do:-)